lib/charms/operator_libs_linux/v0/sysctl.py

```diff
@@ -0,0 +1,229 @@
+        return self.args[0]
+
+
+class SysctlError(Error):
+    """Raised when there's an error running sysctl command."""
+
+
+class SysctlPermissionError(Error):
+    """Raised when there's an error applying values in sysctl."""
+
+
+class ValidationError(Error):
+    """Exception representing value validation error."""
+
+
+class SysctlConfig(Mapping[str, int]):
+    """Represents the state of the config that a charm wants to enforce."""
+
+    def __init__(self, name: str = "") -> None:
+        self.name = name
+
+    def __contains__(self, key: str) -> bool:
+        """Check if key is in config."""
+        return key in self._data
+
+    def __len__(self):
+        """Get size of config."""
+        return len(self._data)
+
+    def __iter__(self):
+        """Iterate over config."""
+        return iter(self._data)
+
+    def __getitem__(self, key: str) -> int:
+        """Get value for key form config."""
+        return self._data[key]
+
+    @property
+    def _data(self) -> Dict[str, int]:
+        """Return applied internal config."""
+        if not self.merged_config_exists:
+            return {}
+
+        with open(SYSCTL_FILENAME, "r") as f:
+            return {param.strip(): int(value.strip())
+                    for line in f.read().splitlines()
+                    if line and not line.startswith('#')
+                    for param, value in [line.split('=')]}
+
+    @property
+    def name(self) -> str:
+        """Name used to create the lib file."""
+        return self._name
+
+    @name.setter
+    def name(self, value: str):
+        if value == "":
+            self._name, *_ = os.getenv("JUJU_UNIT_NAME", "unknown").split("/")
+        else:
+            self._name = value
+
+    @property
+    def charm_filepath(self) -> str:
+        """Name for resulting charm config file."""
+        return f"{SYSCTL_DIRECTORY}/90-juju-{self.name}"
+
+    @property
+    def charm_config_exists(self) -> bool:
+        """Return whether the charm config file exists."""
+        return os.path.exists(self.charm_filepath)
+
+    @property
+    def merged_config_exists(self) -> bool:
+        """Return whether a merged config file exists."""
+        return os.path.exists(SYSCTL_FILENAME)
+
+    def update(self, config: Dict[str, dict]) -> None:
+        """Update sysctl config options with a desired set of config params."""
+        self._parse_config(config)
+        if self.charm_config_exists:
+            return
+        self._create_charm_file()
+
+        conflict = self.validate()
+        if conflict:
+            msg = f"Validation error for keys: {conflict}" 
+            raise ValidationError(msg)
+
+        snapshot = self._create_snapshot()
+        logger.debug(f"Created snapshot for keys: {snapshot}")
+        try:
+            self._apply()
+        except SysctlPermissionError:
+            self._restore_snapshot(snapshot)
+            raise
+        except SysctlError:
+            raise
```

Comment:
I think that we need to remove charm config in both scenarios, so only valid configs are stored. The reason can be explained with simple example:
You already have two charms A and B on machine (two config files) and you tried to deployed third one C, which failed due some conflicts with charm A.  Configs from charm A and B are applied, but configs from all charms are in directory. The C charm is in blocked state (good approach) and you are tied to remove charm A, so config is removed but during marge phase it will failed, since there are still configs for B and C. You will end-up with charm A in error state and charm C in blocked state.

If config file is not there charm A will be removed and you can simply resolved the charm C and it will fine.

Reply:
Yes, I tend to agree. Actually I had a thought in the comment above about this as well. Also, I'm afraid that even if you try to change the configuration of "A" and "B" (that used to work without C), that would start producing exceptions.